### ISSUE TITLE
Fix PytorchLightningPruningCallback for pytorch_lightning>=1.8.0

### ISSUE THREAD

**Issue by wbeardall** [optuna/optuna#4499]
### Motivation

In `pytorch_lightning` >= 1.8.0, the `on_init_start` callback hook is removed (deprecated since v1.6.0). This naturally breaks the `PytorchLightningPruningCallback` for newer versions of Lightning. There are some other issues that arise from architecture changes in newer versions of the `Trainer` object, which I address below.

### Suggestion

This can be fixed by making a few modifications:

1. Change the callback intialization from `on_init_start` to `setup`.
2. Modify the distributed backend check to work with the new `AcceleratorConnector`. This can be done by checking the connector's `strategy` attr.
3. Modify the `should_stop` check to work with the new `AcceleratorConnector`. This can be done by checking the `strategy` attr instead of the `training_type_plugin` attr.

I haven't tested this in DDP scenarios, so it might need a little tweaking, but it seems to work in the simple training case.

```
import warnings

from packaging import version

import optuna
from optuna.storages._cached_storage import _CachedStorage
from optuna.storages._rdb.storage import RDBStorage


# Define key names of `Trial.system_attrs`.
_PRUNED_KEY = "ddp_pl:pruned"
_EPOCH_KEY = "ddp_pl:epoch"


with optuna._imports.try_import() as _imports:
    import pytorch_lightning as pl
    from pytorch_lightning import LightningModule
    from pytorch_lightning import Trainer
    from pytorch_lightning.callbacks import Callback
    from pytorch_lightning.strategies import ParallelStrategy

if not _imports.is_successful():
    Callback = object  # type: ignore  # NOQA
    LightningModule = object  # type: ignore  # NOQA
    Trainer = object  # type: ignore  # NOQA

def _is_ddp_backend(strategy):
    """Check for DDP training strategy.

    This may be too inclusive a check, but I haven't tested it comprehensively.
    Derived from https://github.com/Lightning-AI/lightning/blob/e57f12851fe68ccfb190b368e173d5f4fad67a41/src/lightning/pytorch/trainer/connectors/accelerator_connector.py#L223
    """
    is_ddp_str = isinstance(strategy, str) and "ddp" in strategy
    is_deepspeed_str = isinstance(strategy, str) and "deepspeed" in strategy
    return isinstance(strategy, ParallelStrategy) or is_ddp_str or is_deepspeed_str

if version.parse(pl.__version__) >= version.parse("1.8.0"):

    class PyTorchLightningPruningCallback(Callback):
        """PyTorch Lightning callback to prune unpromising trials.

        See `the example <https://github.com/optuna/optuna-examples/blob/
        main/pytorch/pytorch_lightning_simple.py>`__
        if you want to add a pruning callback which observes accuracy.

        Args:
            trial:
                A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
                objective function.
            monitor:
                An evaluation metric for pruning, e.g., ``val_loss`` or
                ``val_acc``. The metrics are obtained from the returned dictionaries from e.g.
                ``pytorch_lightning.LightningModule.training_step`` or
                ``pytorch_lightning.LightningModule.validation_epoch_end`` and the names thus depend on
                how this dictionary is formatted.

        .. note::
            For the distributed data parallel training, the version of PyTorchLightning needs to be
            higher than or equal to v1.5.0. In addition, :class:`~optuna.study.Study` should be
            instantiated with RDB storage.
        """

        def __init__(self, trial: optuna.trial.Trial, monitor: str) -> None:
            _imports.check()
            super().__init__()

            self._trial = trial
            self.monitor = monitor
            self.is_ddp_backend = False

        def setup(self, trainer: Trainer, pl_module: LightningModule, stage: str) -> None:
            # Fix
            self.is_ddp_backend = _is_ddp_backend(trainer._accelerator_connector.strategy)
            if self.is_ddp_backend:
                if not (
                    isinstance(self._trial.study._storage, _CachedStorage)
                    and isinstance(self._trial.study._storage._backend, RDBStorage)
                ):
                    raise ValueError(
                        "optuna.integration.PyTorchLightningPruningCallback"
                        " supports only optuna.storages.RDBStorage in DDP."
                    )

        def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> None:

            # When the trainer calls `on_validation_end` for sanity check,
            # do not call `trial.report` to avoid calling `trial.report` multiple times
            # at epoch 0. The related page is
            # https://github.com/PyTorchLightning/pytorch-lightning/issues/1391.
            if trainer.sanity_checking:
                return

            epoch = pl_module.current_epoch

            current_score = trainer.callback_metrics.get(self.monitor)
            if current_score is None:
                message = (
                    "The metric '{}' is not in the evaluation logs for pruning. "
                    "Please make sure you set the correct metric name.".format(self.monitor)
                )
                warnings.warn(message)
                return

            should_stop = False
            if trainer.is_global_zero:
                self._trial.report(current_score.item(), step=epoch)
                should_stop = self._trial.should_prune()
            should_stop = trainer.strategy.broadcast(should_stop)
            if not should_stop:
                return

            if not self.is_ddp_backend:
                message = "Trial was pruned at epoch {}.".format(epoch)
                raise optuna.TrialPruned(message)
            else:
                # Stop every DDP process if global rank 0 process decides to stop.
                trainer.should_stop = True
                if trainer.is_global_zero:
                    self._trial.storage.set_trial_system_attr(self._trial._trial_id, _PRUNED_KEY, True)
                    self._trial.storage.set_trial_system_attr(self._trial._trial_id, _EPOCH_KEY, epoch)

        def on_fit_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
            if not self.is_ddp_backend:
                return

            # Because on_validation_end is executed in spawned processes,
            # _trial.report is necessary to update the memory in main process, not to update the RDB.
            _trial_id = self._trial._trial_id
            _study = self._trial.study
            _trial = _study._storage._backend.get_trial(_trial_id)  # type: ignore
            _trial_system_attrs = _study._storage.get_trial_system_attrs(_trial_id)
            is_pruned = _trial_system_attrs.get(_PRUNED_KEY)
            epoch = _trial_system_attrs.get(_EPOCH_KEY)
            intermediate_values = _trial.intermediate_values
            for step, value in intermediate_values.items():
                self._trial.report(value, step=step)

            if is_pruned:
                message = "Trial was pruned at epoch {}.".format(epoch)
                raise optuna.TrialPruned(message)

else:
    PyTorchLightningPruningCallback = optuna.integration.PyTorchLightningPruningCallback
```

**Comment by Alnusjaponica** [comment]
We appreciate your detailed report and effort to find out a solution.
Actually, the problem is already discussed in #3418 and resolved by #4322. Although this PR temporary stop supporting DDP, #4384 is going to follow it up.

**Comment by elephantoid** [comment]
I have a same issue to using both of Optuna and PL.
I'll test DDP with using setup function. Thanks to report and sharing!

**Comment by Alnusjaponica** [comment]
@elephantoid 
Thank you for your comment.
Although the PR is not merged into the `master` branch, it is also possible to install optuna by specifying fork and branch (see [pip document](https://pip.pypa.io/en/stable/user_guide/#:~:text=Requirements%20files%20are%20used%20to%20override,file%20with%20a%20line%20like%20so%3A) for more details). In this case, run the following command:
```
pip install git+ssh://git@github.com/Alnusjaponica/optuna.git@fix-checks-integration-pytorch-lightning-followup
```
Also, the example code in optuna/optuna-examples will be updated according to this change. See optuna/optuna-examples#179 for more details.

**Comment by Alnusjaponica** [comment]
Hi @wbeardall, now that https://github.com/optuna/optuna/pull/4384 is merged and `PytorchLightningPruningCallback` should work with `PyTorch Lightning` >= 1.8. Could you confirm whether this update resolves your problem?

**Comment by hkang-vuno** [comment]
From what I can tell, optuna 3.1.1 still contains "on_init_start". Would the pip install command above still work? 

Thank you.

**Comment by Alnusjaponica** [comment]
As written above, this issue exists under `v3.1.1` and solved on the master. If you would like use `PytorchLightningPruningCallback` before `v3.1.2` release, try installing master branch as follows:
```
pip install git+https://git@github.com/optuna/optuna.git@master
```

**Comment by Alnusjaponica** [comment]
This problem is fixed in v3.2.
